### PR TITLE
Add support for insert(elem)

### PR DESCRIPTION
**Pull request by zaporter**
The ability to insert into this structure can be very important if you have a read-heavy but right-necessary workload where rebuilding the whole data structure for every write is infeasible.

I have tried to implement this function so that it is easy to understand and maintain. It might not be as fast as it could be (that is outside my area of expertise). 

Please let me know if you have intentionally avoided adding this function for whatever reason or if you think that I should be using a different data structure. 

I also refactored some of the tests. I did this just in case you want to add other methods in the future, it seemed like the right thing to do. Let me know if you have any opinions on that. 

Thank you for your cool library :heart:

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -955,6 +1085,43 @@ mod tests {
         assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
         assert_eq!(lapper.intervals.len(), lapper.starts.len())
 
+    }
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion
    }
    
```

File: src/lib.rs [hunk 2]
```diff
@@ -955,6 +1085,43 @@ mod tests {
...
+            &Iv{start: 50, stop: 55, val: 0},
+            &Iv{start: 60, stop: 65, val: 0},
+            &Iv{start: 68, stop: 120, val: 0}, // max_len = 50
+        ];
+        assert_eq!(lapper.intervals.len(), lapper.starts.len());
+        lapper.merge_overlaps();
+        assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
+        assert_eq!(lapper.intervals.len(), lapper.starts.len())
+
+    }
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion
    }

```

File: src/lib.rs [hunk 3]
```diff
@@ -955,6 +1085,43 @@ mod tests {
...
+            &Iv{start: 40, stop: 45, val: 0},
+            &Iv{start: 50, stop: 55, val: 0},
+            &Iv{start: 60, stop: 65, val: 0},
+            &Iv{start: 68, stop: 120, val: 0}, // max_len = 50
+        ];
+        assert_eq!(lapper.intervals.len(), lapper.starts.len());
+        lapper.merge_overlaps();
+        assert_eq!(expected, lapper.iter().collect::<Vec<&Iv>>());
+        assert_eq!(lapper.intervals.len(), lapper.starts.len())
+
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion
```

File: src/lib.rs [hunk 4]
```diff
@@ -996,6 +1163,18 @@ mod tests {
         lapper.set_cov();
         assert_eq!(lapper.cov(), 50);
     }
+    #[test]
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    #[test]
```

File: src/lib.rs [hunk 5]
```diff
@@ -939,6 +1048,27 @@ mod tests {
         );
         assert_eq!(lapper.count(8, 20), 2);
     }
+    #[test]
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    #[test]
```

File: src/lib.rs [hunk 6]
```diff
@@ -872,6 +967,20 @@ mod tests {
         assert_eq!(Some(&expected), lapper.seek(15, 25, &mut cursor).next());
         assert_eq!(lapper.find(15, 25).count(), lapper.count(15, 25));
     }
+    // Test that a query that overlaps the start of an interval returns that interval
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    // Test that a query that overlaps the start of an interval returns that interval
```

File: src/lib.rs [hunk 7]
```diff
@@ -857,6 +943,15 @@ mod tests {
         assert_eq!(None, lapper.seek(30, 35, &mut cursor).next());
         assert_eq!(lapper.find(30, 35).count(), lapper.count(30, 35));
     }
+    // Test that a query start that hits an interval end returns no interval
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    // Test that a query start that hits an interval end returns no interval
```

File: src/lib.rs [hunk 8]
```diff
@@ -847,6 +924,15 @@ mod tests {
         assert_eq!(None, lapper.seek(15, 20, &mut cursor).next());
         assert_eq!(lapper.find(15, 20).count(), lapper.count(15, 20));
     }
+    // Test that a query stop that hits an interval start returns no interval
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    // Test that a query stop that hits an interval start returns no interval
```

File: src/lib.rs [hunk 9]
```diff
@@ -826,8 +893,18 @@ mod tests {
...
             Iv{start: 68, stop: 71, val: 0}, // overlap start
             Iv{start: 70, stop: 75, val: 0},
-        ];
-        Lapper::new(data)
+        ]
+    }
+
+    fn setup_badlapper() -> Lapper<usize, u32> {
+        Lapper::new(get_data_badlapper())
+    }
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion
    }

```

File: src/lib.rs [hunk 10]
```diff
@@ -792,30 +840,49 @@ mod tests {
...
+        Lapper::new(get_data_overlapping())
+    }
+    fn setup_overlapping_insert() -> Lapper<usize, u32> {
+        let mut lapper = Lapper::new(vec![]);
+        for elem in get_data_overlapping() {
+            lapper.insert(elem);
+        }
+        lapper
+
+    }
```
Comment (sstadick) [inline src/lib.rs:963]:
```suggestion
    }

```

File: src/lib.rs [hunk 11]
```diff
@@ -792,30 +840,49 @@ mod tests {
...
-            .collect();
-        Lapper::new(data)
+            .collect()
     }
-    fn setup_badlapper() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = vec![
+    fn setup_overlapping() -> Lapper<usize, u32> {
+        Lapper::new(get_data_overlapping())
+    }
+    fn setup_overlapping_insert() -> Lapper<usize, u32> {
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    fn setup_overlapping_insert() -> Lapper<usize, u32> {
```

File: src/lib.rs [hunk 12]
```diff
@@ -792,30 +840,49 @@ mod tests {
...
-            .collect();
-        Lapper::new(data)
+            .collect()
     }
-    fn setup_overlapping() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = (0..100)
+    fn setup_nonoverlapping() -> Lapper<usize, u32> {
+        Lapper::new(get_data_nonoverlapping())
+    }
+    fn setup_nonoverlapping_insert() -> Lapper<usize, u32> {
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    fn setup_nonoverlapping_insert() -> Lapper<usize, u32> {
```

File: src/lib.rs [hunk 13]
```diff
@@ -792,30 +840,49 @@ mod tests {
...
                 stop: x + 10,
                 val: 0,
             })
-            .collect();
-        Lapper::new(data)
+            .collect()
     }
-    fn setup_overlapping() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = (0..100)
+    fn setup_nonoverlapping() -> Lapper<usize, u32> {
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    fn setup_nonoverlapping() -> Lapper<usize, u32> {
```

File: src/lib.rs [hunk 14]
```diff
@@ -792,30 +840,49 @@ mod tests {
...
+        Lapper::new(get_data_nonoverlapping())
+    }
+    fn setup_nonoverlapping_insert() -> Lapper<usize, u32> {
+        let mut lapper = Lapper::new(vec![]);
+        for elem in get_data_nonoverlapping() {
+            lapper.insert(elem);
+        }
+        lapper
+    }
+    fn get_data_overlapping() -> Vec<Iv> {
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    fn get_data_overlapping() -> Vec<Iv> {
```

File: src/lib.rs [hunk 15]
```diff
@@ -792,30 +840,49 @@ mod tests {
     use super::*;
 
     type Iv = Interval<usize, u32>;
-    fn setup_nonoverlapping() -> Lapper<usize, u32> {
-        let data: Vec<Iv> = (0..100)
+    fn get_data_nonoverlapping() -> Vec<Iv> {
```
Comment (sstadick) [inline src/lib.rs]:
```suggestion

    fn get_data_nonoverlapping() -> Vec<Iv> {
```

File: src/lib.rs [hunk 16]
```diff
@@ -373,16 +410,27 @@ where
...
 
     #[inline]
-    pub fn bsearch_seq(key: I, elems: &[I]) -> usize {
-        if elems[0] > key {
+    pub fn bsearch_seq<K>(key: K, elems: &[K]) -> usize 
+        where K: PartialEq+PartialOrd
+    {
+        Self::bsearch_seq_ref(&key, elems)
+    }
+    #[inline]
```
Comment (sstadick) [inline src/lib.rs:420]:
```suggestion

    #[inline]
```

**Review by sstadick — APPROVED** [review]
@zaporter Thanks for the contribution! This is excellent, I had just a few nit-picky changes that I've proposed (apologies if you were following formatting that I already had in place, I've evolved!)

Assuming this passes all tests I'll merge it once the changes are in and cut a release. Thank you!

**Comment by sstadick** [comment]
@zaporter - lookslike it's just failing the cargofmt lint. Could you run that on your branch and push to here? I think we should be in a good spot after that.

**Comment by sstadick** [comment]
I only never added an insert because I didn't have that use case and I knew it would be slooooow. But I have nothing against having it available with the appropriate caveats in place in the docs 👍

**Comment by zaporter** [comment]
@sstadick Thanks for the formatting advice! I forgot to run a format check before submitting this :facepalm: . 

Let me know if you have any further comments, but I think I addressed everything.
 
Re: not implemented: Yeah, an insert is slow, but hopefully this will still be faster than other data structures for a read heavy workload. I need to do some benchmarking... If I get any interesting numbers I might submit them just so other people know. 

Again, thanks for making this!

**Comment by zaporter** [comment]
I woke up this morning and realized that my tests weren't testing what they should be. Just rewrote the tests. This should be a no-brainer switch. 

All of the tests pass and it guarantees that creating from ::new() or creating from ::insert() create the same Lapper. 

Ready to merge after your approval.

**Review by sstadick — APPROVED** [review]

**Comment by sstadick** [comment]
@zaporter This all looks great, thank you again! I'll make the release tomorrow 👍 and I haven't done it by tomorrow feel free to ping me.

**Comment by sstadick** [comment]
@zaporter released, see v1.1.0. Thanks again!

**Pull request merged**